### DISCUSSION TITLE
Vídeo 03 - Configuração

### DISCUSSION THREAD

**Discussion by thpoiani** [thpoiani/lab-quarkus#4] [General]
Obrigado por acompanhar mais um vídeo!
Dessa vez comentamos sobre padrões de arquitetura e iniciamos a estrutura de um projeto com Quarkus.

Posso citar algumas vantagens ao utilizar um padrões de arquitetura: modularidade, testabilidade, redução de acoplamento, ...

Você havia trabalhado com algum desses padrões ou com algum outro estilo (_Model-View-Controller_, _Model-View-Presenter_, _Model-View-ViewModel_, _Event Driven Architecture_, ...)?

Estou curioso... me conte sobre a sua primeira impressão com Quarkus durante esse vídeo 😊

**Comment by didifive** [comment]
Até então trabalhei mais com o estilo _Model-View-Controller_
Já tinha visto alguns conceitos, mas vou aproveitar este projeto para conhecer melhor praticando com arquitetura limpa.

Estou aproveitando também para conhecer o Quarkus, ele destaca bastante o uso com GraalVM e com containers. O bom é que ainda permite usar um banner no carregamento semelhante ao spring, até inclui um no fork que fiz.

**Comment by joseeasantos** [comment]
Olá! Achei bem interessante o modelo cebola e limpa. Já trabalhei com MVC. Me chama a atenção o modelo hexagonal por ter fronteiras muito bem definidas. Creio que o uso em si depende do objetivo de cada projeto.
Neste curso, tem sido o meu primeiro contato com Quarkus. Achei muito parecido com Node.JS executado via NPM, onde podemos alterar o projeto e visualizar em tempo real a execução da alteração. Particularmente, creio que tem tudo para avançar e pretendo aprofundar. Caso seja possível @thpoiani , pode dar algumas dicas de continuidade do conteúdo?

**Reply by thpoiani** [reply]
Olá @joseeasantos, no vídeo final há uma série de sugestões e desafios para continuidade, como por exemplo:
- (Developer) Implemente alterações no código para tornar as aplicações mais reactivas/assíncrona
  - https://quarkus.io/guides/resteasy-reactive#asyncreactive-support
- (DevOps) Provisione uma réplica de leitura do banco de dados, habilitando múltiplas conexões nos repositories
  - https://mariadb.com/kb/en/setting-up-replication https://quarkus.io/guides/hibernate-orm#multiple-persistence-units
- (Data Scientist) Modele um workflow para Análise Preditiva do resultado da eleição

**Comment by ale-fuentes-ar** [comment]
## Finalizando a aula 3 | Configuração
![](https://img.shields.io/badge/by-Alejandro.Fuentes-informational?style=flat-square&logoColor=white&color=cdcdcd)

Boa noite team...e boa noite @thpoiani .. novamente grato por compartir e pela sua atenção.

Eu inicie o processo de configuração.
Eu so não cosigo fazer o processo de "attach proccess". Após de executar `quarkus dev` e após de esperar que apressente a porta em execução (no caso do exemplo a 5005), a selecionar o menu  Run / Attach to process, não aparece nada na lista.


> OBS:
> Eu estou usando Intellij  IDEA como IDE
> OS Windows 11
> ambiente foi criado no WSL.


Caso tiver alguma dica, ou conteudo para leitura sobre o assunto para corregir, eu ficaria grato se puder compartilhar.
Otima noite a todos !!

**Reply by didifive** [reply]
Olá, @ale-fuentes-ar, eu utilizei Windows 10 e WSL. O attach process igual o @thpoiani utilizou não funcionou comigo também. 
No WSL onde eu tinha o recurso do docker eu executei o quarkus dev para ele subir aplicação pelo WSL também (ja que docker só existia no WSL, tinha que subir a aplicação por ele para funcionar o testcontainers)
Depois no IntelliJ eu configurei um run semelhante ao que tem no item 3 do tutorial: https://www.baeldung.com/intellij-remote-debugging

Apanhei muito com isso de Windows 10 e WSL, mas deu para aprender bastante. 
Única coisa que aconteceu e não resolvi foi o live reload do quarkus dev. Quando alterava algo tinha que reiniciar a aplicação.

**Reply by ale-fuentes-ar** [reply]
Estimado @didifive .. fico super grato pela resposta, ire dar uma olhada no links que você compartilhou.

Eu estou apanhando com ambientes, queria saber qual seria a forma de criar um ambiente robusto (falo em sentido ou estilo *nao tem erro*) para poder practicar ..

**Reply by thpoiani** [reply]
Obrigado pelo suporte @didifive.
@ale-fuentes-ar esse parece ser então um problema mais relacionado a como o sistema operacional gerencia os containers.

Eu não consigo reproduzir esse cenário para dar uma resposta mais objetiva, mas segue algumas ideias que podem ajudar nesse caso:
- veja essa documentação sobre **remote development**: https://quarkus.io/guides/maven-tooling#remote-development-mode
- não sei se isso pode ser um problema específico do Docker no Windows... considere outro container runtime, como https://github.com/containerd/containerd por exemplo

**Reply by thpoiani** [reply]
> queria saber qual seria a forma de criar um ambiente robusto

@ale-fuentes-ar, como toda a possível resposta de Engenharia de Software: **isso depende**.
O que você quer dizer com **_robusto_**?
Seria _high availability_? Ou talvez _reliability_? Pode até ser _durability_? Ou _resiliency_? Talvez _fault tolerance_? _Scalability_? _Elasticity_? _Security_? ...
Existe uma série de características em arquitetura de software para tentar responder isso. E claro, isso também depende muito do seu produto e do comportamento do usuário. E essa decisão pode levar a uma escolha arquitetural e de infraestrutura muito diferente da outra.

**Reply by ale-fuentes-ar** [reply]
Estimado @thpoiani .. fico muio grato pelas suas respostas...

1- ire tentar sua dica sobre gerenciamento de containers (*estimo que deve ser o escenario do SO, ja que estou em WSL, em W11, etc)
2- desculpe pela espressão vaga ou falta de uma melhor expressão. A verdade o que quiz dizer e sobre como criar um entorno de desenvolvimento, sem que o SO atrapalhe ... mas eu compreendi, precisso aprimorar mais o lado de configurar e preparar ambientes de desenvolvimento.

**Comment by mauromortoni** [comment]
Sim, trabalhei com Model-View-Controller e aguardo projetos desafiadores para colocar em prática o conhecimento adquirido.

Minha primeira impressão sobre o Quarkus: é uma plataforma inovadora que oferece uma interface web rica em funcionalidades práticas para desenvolvedores. Sua arquitetura moderna permite uma integração fluida com diversos frameworks, o que simplifica a curva de aprendizado e acelera o desenvolvimento de aplicações. Com recursos como suporte a programação reativa, otimizações para GraalVM e uma abordagem centrada no Kubernetes, o Quarkus capacita equipes a criar soluções ágeis e escaláveis, tornando o processo de desenvolvimento mais eficiente e produtivo.
